x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -5,75 +5,350 @@
+            numConcurrentReads++;
+            long from = lastRequestedSeqno + 1;
+            LOGGER.trace("{}[{}] peek read [{}]", params.getFollowShardId(), numConcurrentReads, from);
+            sendShardChangesRequest(from, maxReadSize, null);
+        }
+    }
+
+    private boolean hasReadBudget() {
+        assert Thread.holdsLock(this);
+        if (numConcurrentReads >= params.getMaxConcurrentReads()) {
+            LOGGER.trace("{} no new reads, maximum number of concurrent reads have been reached [{}]",
+                params.getFollowShardId(), numConcurrentReads);
+            return false;
+        }
+        if (buffer.size() > params.getMaxBufferSize()) {
+            LOGGER.trace("{} no new reads, buffer limit has been reached [{}]", params.getFollowShardId(), buffer.size());
+            return false;
+        }
+        return true;
+    }
+
+    private synchronized void coordinateWrites() {
+        while (hasWriteBudget() && buffer.isEmpty() == false) {
+            if (numConcurrentWrites >= params.getMaxConcurrentWrites()) {
+                LOGGER.trace("{} maximum number of concurrent writes have been reached [{}]",
+                    params.getFollowShardId(), numConcurrentWrites);
+                break;
+            }
+            Translog.Operation[] ops = new Translog.Operation[Math.min(params.getMaxWriteSize(), buffer.size())];
+            for (int i = 0; i < ops.length; i++) {
+                ops[i] = buffer.remove();
+            }
+            numConcurrentWrites++;
+            LOGGER.trace("{}[{}] write [{}/{}] [{}]", params.getFollowShardId(), numConcurrentWrites, ops[0].seqNo(),
+                ops[ops.length - 1].seqNo(), ops.length);
+            sendBulkShardOperationsRequest(ops);
+        }
+    }
+
+    private boolean hasWriteBudget() {
+        assert Thread.holdsLock(this);
+        if (numConcurrentWrites >= params.getMaxConcurrentWrites()) {
+            LOGGER.trace("{} maximum number of concurrent writes have been reached [{}]",
+                params.getFollowShardId(), numConcurrentWrites);
+            return false;
+        }
+        return true;
+    }
+
+
+    private void sendShardChangesRequest(long from, long size, Long targetSeqNo) {
+        innerSendShardChangesRequest(from, size,
+            response -> {
+                retryCounter.set(0);
+                handleResponse(from, size, targetSeqNo, response);
+            },
+            e -> handleFailure(e, () -> sendShardChangesRequest(from, size, targetSeqNo)));
     }
 
+    private void handleResponse(long from, long size, Long targetSeqNo, ShardChangesAction.Response response) {
+        maybeUpdateMapping(response.getIndexMetadataVersion(), () -> {
+            synchronized (ShardFollowNodeTask.this) {
+                globalCheckpoint = Math.max(globalCheckpoint, response.getLeaderGlobalCheckpoint());
+                if (response.getOperations().length == 0) {
+                    if (targetSeqNo != null) {
```

Comment:
if we change this to never be null, I think we can share the logic here and make it simpler. Something like:

1) Add everything we get to the buffer
2) coordinatewrites (always)
3) If max retrieved seq!= requiredMaxSeq, send another request
4) else if ops.length == 0 schedule a future coordinate read
5) else if coordinate read now.